nonlinear_solid_mechanics/include/coupling_functions.h

```diff
@@ -0,0 +1,370 @@
+       *             individual configuration and preCICE determines it
+       *             automatically. In many cases, this data will just represent
+       *             your initial condition.
+       * @param[out] precice_to_deal
+       */
+      void
+      initialize_precice(const DoFHandler<dim> &dof_handler,
+                         const VectorType &     deal_to_precice,
+                         VectorType &           precice_to_deal);
+
+      /**
+       * @brief      advance_precice
+       *
+       * @param[in]  deal_to_precice
+       * @param[out] precice_to_deal
+       * @param[in]  computedTimestepLength
+       */
+      void
+      advance_precice(const VectorType &deal_to_precice,
+                      VectorType &      precice_to_deal,
+                      const double      computedTimestepLength);
+
+      /**
+       * @brief      Saves time dependent variables in case of an implicit coupling
+       *
+       * @param[in]  state_variables vector containing all variables to store
+       *
+       * @note       This function only makes sense, if it is used with
+       *             @p reload_old_state. Therefore, the order, in which the
+       *             variables are passed into the vector must be the same for
+       *             both functions.
+       * @note       The absolute time has no impact on the computation, but on the output.
+       *             Therefore, we call here in the @p Time class a method to store the
+       *             current time and reload it later. This is necessary for
+       *             subcycling.
+       */
+      void
+      save_current_state(const std::vector<VectorType *> &state_variables,
+                         Time &                           time_class);
+
+      /**
+       * @brief      Reloads the previously stored variables in case of an implicit
+       *             coupling
+       *
+       * @param[out] state_variables vector containing all variables to reload
+       *
+       * @note       This function only makes sense, if the state variables have been
+       *             stored by calling @p save_current_state. Therefore, the order, in
+       *             which the variables are passed into the vector must be the
+       *             same for both functions.
+       */
+      void
+      reload_old_state(std::vector<VectorType *> &state_variables,
+                       Time &                     time_class);
+
+      /**
+       * @brief public precice solverinterface
+       */
+
+      precice::SolverInterface precice;
+
+    private:
+      // preCICE related initializations
+      const unsigned int deal_boundary_id;
+      const bool         enable_precice;
+      const std::string  mesh_name;
+      const std::string  read_data_name;
+      const std::string  write_data_name;
+
+      int precice_mesh_id;
+      int n_interface_nodes;
+      int ptd_data_id;
+      int dtp_data_id;
```

Comment:
Yes, I thought already, that you're going to complain about it. I honestly speaking don't like read and write as names, since it depends on your perspective. But that's consistent with preCICE and I will change it.